### PR TITLE
region_syncer: reduce saveKV of client

### PR DESCRIPTION
**Pull request by HunDunDM**
### What is changed and how it works?

* By introducing part of the heartbeat processing logic in the client of `region_syncer`, unnecessary `saveKV` operations are reduced.

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
Improved the performance of synchronizing Region information between PDs.
```

### CODE REVIEW

File: server/region_syncer/server.go [hunk 1]
```diff
@@ -210,9 +212,11 @@ func (s *RegionSyncer) syncHistoryRegion(request *pdpb.SyncRegionRequest, stream
 				lastIndex += len(metas)
 				if err := stream.Send(resp); err != nil {
 					log.Error("failed to send sync region response", errs.ZapError(errs.ErrGRPCSend, err))
+					return err
 				}
 				metas = metas[:0]
 				stats = stats[:0]
+				leaders = leaders[:0]
```
Comment (HunDunDM) [inline server/region_syncer/server.go:219]: Will this cause the new PD instance to have incorrect leader information? @rleungx

File: server/core/region.go [hunk 2]
```diff
@@ -460,6 +462,95 @@ func (r *RegionInfo) GetReplicationStatus() *replication_modepb.RegionReplicatio
 	return r.replicationStatus
 }
 
+// RegionGuideFunc is a function that determines which follow-up operations need to be performed based on the origin
+// and new region information.
+type RegionGuideFunc func(region, origin *RegionInfo) (isNew, saveKV, saveCache, needSync bool)
+
+// GenerateRegionGuideFunc is used to generate a RegionGuideFunc. Control the log output by specifying the log function.
+// nil means do not print the log.
+func GenerateRegionGuideFunc(debug, info func(msg string, fields ...zap.Field)) RegionGuideFunc {
```
Comment (disksing) [inline server/core/region.go:471]: Maybe `GenerateRegionGuideFunc(enableLog bool)` is better. Or `GuideRegion(region, origin *RegionInfo, enableLog bool) (isNew, saveKV, saveCache, needSync bool)` directly.

**Review by disksing — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**